drivers/dac/dac_ad569x.c

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright (c) 2024 Jan Kubiznak <jan.kubiznak@deveritec.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdint.h>
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/dac.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/byteorder.h>
+
+LOG_MODULE_REGISTER(dac_ad569x, CONFIG_DAC_LOG_LEVEL);
+
+#define AD569X_CTRL_GAIN(x)  FIELD_PREP(BIT(11), x)
+#define AD569X_CTRL_REF(x)   FIELD_PREP(BIT(12), x)
+#define AD569X_CTRL_PD(x)    FIELD_PREP(BIT_MASK(2) << 13, x)
+#define AD569X_CTRL_RESET(x) FIELD_PREP(BIT(15), x)
+
+#define AD569X_CMD_WRITE            (0x10)
+#define AD569X_CMD_UPDATE           (0x20)
+#define AD569X_CMD_WRITE_AND_UPDATE (0x30)
+#define AD569X_CMD_CONFIGURE        (0x40)
+
+#define AD569X_CTRL_NO_RESET      (0x00)
+#define AD569X_CTRL_PERFORM_RESET (0x01)
+
+struct ad569x_config {
+	struct i2c_dt_spec bus;
+	uint8_t resolution;
+	uint8_t gain;
+	uint8_t voltage_reference;
+	uint8_t power_down_mode;
+};
+
+static int ad569x_write(const struct device *dev, uint8_t command, uint16_t value)
+{
+	const struct ad569x_config *config = dev->config;
+
+	uint8_t tx_data[3];
+
+	tx_data[0] = command;
+	sys_put_be16(value, tx_data + 1);
+
+	return i2c_write_dt(&config->bus, tx_data, sizeof(tx_data));
+}
+
+static int ad569x_read(const struct device *dev, uint16_t *value)
+{
+	const struct ad569x_config *config = dev->config;
+
+	uint8_t rx_data[2];
+	int ret;
+
+	ret = i2c_read_dt(&config->bus, rx_data, sizeof(rx_data));
+	if (ret != 0) {
+		return ret;
+	}
+
+	*value = sys_get_be16(rx_data);
+
+	return ret;
+}
+
+static int ad569x_channel_setup(const struct device *dev, const struct dac_channel_cfg *channel_cfg)
+{
+	const struct ad569x_config *config = dev->config;
+
+	if (channel_cfg->channel_id > 0) {
+		LOG_ERR("invalid channel %d", channel_cfg->channel_id);
+		return -EINVAL;
+	}
+
+	if (channel_cfg->resolution != config->resolution) {
+		LOG_ERR("invalid resolution %d", channel_cfg->resolution);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int ad569x_sw_reset(const struct device *dev)
+{
+	uint16_t reg = AD569X_CTRL_RESET(AD569X_CTRL_PERFORM_RESET);
+
+	LOG_DBG("reset %s", dev->name);
+
+	/* Ignore return value, since device gives NAK after receiving RESET request */
+	ad569x_write(dev, AD569X_CMD_CONFIGURE, reg);
+
+	/* Check that DAC output is reset */
+	int ret = ad569x_read(dev, &reg);
+
```

Comment:
nit: it's good practice to keep the check after the operation, drop the blank line here, then compliance will complain that you need a blank line after variable declaration so declare the variable on its own at the beginning of the function